### DISCUSSION TITLE
What is the best way to compare two objects within a `PyIterator` (`PyAny`)

### DISCUSSION THREAD

**Discussion by tgross35** [PyO3/pyo3#2502] [Questions] [answered]
Hello all,

I have a handful of existing functions that require `PartialEq` on items in an iterator. What is a good way to transform from a `PyIterator` to this? It seems like this type has `PyAny` as its item - and even though this has an `eq` function, it can't be directly compared.

I am thinking that two options are either to try to downcast into various python types that might be comparable, or to implement `PartialEq` by unwrapping the `eq` and `ne` methods on `PyAny`, but it seems like there must be a reason those options aren't available by default.

**Comment by davidhewitt** [comment, marked as answer]
> implement `PartialEq` by unwrapping the `eq` and `ne` methods on PyAny, 

You probably want to wrap your `PyAny` objects in a newtype which does exactly this, or alternatively could return `None` from `PartialEq::cmp` if there's an exception raised during comparison. For now we're undecided what the correct error handling strategy is if PyO3 were to provide a `PartialEq` implementation.

Some brief history - `PyAny` used to implement `PartialEq` using pointer comparison (i.e. Python `is`), which was almost certainly a footgun. We removed that implementation a couple of versions ago, and now in a cooling off period to take stock before we consider if we add a more correct implementation back.

**Reply by tgross35** [reply]
Interesting, thank you for the context!

On a similar note, is there a good way to get a `impl DoubleEndedIterator` from a `PySequence`?

**Reply by davidhewitt** [reply]
I think you'd need to again create a newtype `DoubleEndedIterator<'a>(&'a PySequence)` and implement it yourself. Should be straightforward enough; only thing to bear in mind is that because Python objects are freely mutable you might want to check length on each iterator (& panic if changed?)
